gateway_tests/process/test_subscriptions.py

```diff
@@ -0,0 +1,251 @@
+            )
+            return
+
+        if all(
+            isinstance(value1, (int, float))
+            and math.isnan(value1)
+            and value2 == 0.0
+            for key, value1, value2 in conftest.find_differences(
+                gateway_events[0], gateway_events[1]
+            )
+        ):
+            if strict:
+                raise RuntimeError(f"NaN event and then 0.0 event:\n{differences}")
+                return
+
+            logger.warning(
+                "Partial passed test - gateway behaves slightly differently.  "
+                "Gateway duplicated sub callback with NaN then 0.0 for some "
+                "values."
+            )
+            return
+        else:
+            raise RuntimeError(f"Differences in events:\n{differences}")
+
+    assert len(gateway_events) == len(ioc_events), (
+        f"Gateway events = {len(gateway_events)}, "
+        f"but IOC events = {len(ioc_events)}."
+    )
+
+
+@pytest.mark.parametrize(
+    "pvname, values",
+    [
+        pytest.param("HUGO:AI", [0.2, 1.2]),
+        pytest.param("HUGO:ENUM", [1, 2]),
+        pytest.param("enumtest", [1, 2]),
+        pytest.param("gwcachetest", [-20, 0, 20]),
+        pytest.param("passive0", [1, 21]),
+        pytest.param("passiveADEL", [1, 20]),
+        pytest.param("passiveADELALRM", [1, 20]),
+        pytest.param("passiveALRM", [1, 5, 10]),
+        pytest.param("passiveMBBI", [1, 2]),
+        pytest.param("passivelongin", [1, 2]),
+        pytest.param("bigpassivewaveform", [[1, 2, 3], [4, 5, 6]]),
+        # pytest.param("fillingaai", []),
+        # pytest.param("fillingaao", []),
+        # pytest.param("fillingcompress", []),
+        # pytest.param("fillingwaveform", []),
+        # pytest.param("passivewaveform", []),
+    ]
+)
+@forms
+@value_masks
+@conftest.standard_test_environment_decorator
+def test_subscription_with_put(pvname: str, mask: int, form: str, values: list[Any]):
+    """
+    Putting a value to the IOC and compare subscription events.
+
+    For the provided pv name, mask, and form (ctrl/time), do we receive the same
+    subscription updates after putting values to the IOC?
+    """
+    gateway_events = []
+    ioc_events = []
+
+    def on_change(event_list, pvname=None, chid=None, **kwargs):
+        event_list.append(kwargs)
+
+    with conftest.ca_subscription_pair(
+        pvname,
+        ioc_callback=functools.partial(on_change, ioc_events),
+        gateway_callback=functools.partial(on_change, gateway_events),
+        form=form,
+        mask=mask,
+    ) as (ioc_ch, gateway_ch):
+        _wait_event(gateway_events, ioc_events, count=1, timeout=0.2)
+        # Throw away initial events; we care what happens from now on
+        del gateway_events[:]
+        del ioc_events[:]
+        for value in values:
+            ca.put(ioc_ch, value)
+            _wait_event(gateway_events, ioc_events)
+        time.sleep(0.1)
```

Comment:
review note: this sleep appears to be functional: "wait for 1 event, and then wait a little longer"

Reply:
or rather, it's done at the end of the loop- I'm not sure the intention then

Reply:
Sorry about this one - I was trying to figure out how we're getting mismatched events. It now appears that the gateway sees twice the amount of events as a similar subscription to the IOC.